ompi/mca/coll/base/coll_base_util.c

```diff
@@ -103,3 +103,165 @@ int ompi_rounddown(int num, int factor)
+    struct ompi_coll_base_nbc_request_t *request = (ompi_coll_base_nbc_request_t *)req;
+    int rc = OMPI_SUCCESS;
+    assert (NULL != request);
+    if (NULL != request->cb.req_complete_cb) {
+        rc = request->cb.req_complete_cb(request->req_complete_cb_data);
+    }
+    release_objs_callback(request);
+    return rc;
+}
+
+static int free_objs_callback(struct ompi_request_t **rptr) {
+    struct ompi_coll_base_nbc_request_t *request = *(ompi_coll_base_nbc_request_t **)rptr;
+    int rc = OMPI_SUCCESS;
+    if (NULL != request->cb.req_free) {
+        rc = request->cb.req_free(rptr);
+    }
+    release_objs_callback(request);
+    return rc;
+}
+
+int ompi_coll_base_retain_op( ompi_request_t *req, ompi_op_t *op,
+                              ompi_datatype_t *type) {
+    ompi_coll_base_nbc_request_t *request = (ompi_coll_base_nbc_request_t *)req;
+    bool retain = false;
+    if (!ompi_op_is_intrinsic(op)) {
+        OBJ_RETAIN(op);
+        request->data.op.op = op;
+        retain = true;
+    }
+    if (!ompi_datatype_is_predefined(type)) {
+        OBJ_RETAIN(type);
+        request->data.op.datatype = type;
+        retain = true;
+    }
+    if (OPAL_UNLIKELY(retain)) {
+        if (req->req_persistent) {
+            request->cb.req_free = req->req_free;
+            req->req_free = free_objs_callback;
+        } else {
+            request->cb.req_complete_cb = req->req_complete_cb;
+            request->req_complete_cb_data = req->req_complete_cb_data;
+            req->req_complete_cb = complete_objs_callback;
+            req->req_complete_cb_data = request;
+        }
+    }
+    return OMPI_SUCCESS;
+}
+
+int ompi_coll_base_retain_datatypes( ompi_request_t *req, ompi_datatype_t *stype,
+                                     ompi_datatype_t *rtype) {
+    ompi_coll_base_nbc_request_t *request = (ompi_coll_base_nbc_request_t *)req;
+    bool retain = false;
+    if (NULL != stype && !ompi_datatype_is_predefined(stype)) {
+        OBJ_RETAIN(stype);
+        request->data.types.stype = stype;
+        retain = true;
+    }
+    if (NULL != rtype && !ompi_datatype_is_predefined(rtype)) {
+        OBJ_RETAIN(rtype);
+        request->data.types.rtype = rtype;
+        retain = true;
+    }
+    if (OPAL_UNLIKELY(retain)) {
+        if (req->req_persistent) {
+            request->cb.req_free = req->req_free;
+            req->req_free = free_objs_callback;
+        } else {
+            request->cb.req_complete_cb = req->req_complete_cb;
+            request->req_complete_cb_data = req->req_complete_cb_data;
+            req->req_complete_cb = complete_objs_callback;
+            req->req_complete_cb_data = request;
+        }
+    }
+    return OMPI_SUCCESS;
+}
+
+static void release_vecs_callback(ompi_coll_base_nbc_request_t *request) {
+    ompi_communicator_t *comm = request->super.req_mpi_object.comm;
+    int count = OMPI_COMM_IS_INTER(comm)?ompi_comm_remote_size(comm):ompi_comm_size(comm);
```

Comment:
How does this work for persistent neighbor collectives ?

Reply:
I guess you meant non blocking neighbor alltoallw (regardless this is a persistent operation or not), right ? in this case, this is a good catch I fixed in the latest commit.